src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Completion/DirectiveAttributeCompletionItemProvider.cs

```diff
@@ -140,15 +154,42 @@ internal static ImmutableArray<RazorCompletionItem> GetAttributeCompletions(
 
             var razorCompletionItem = RazorCompletionItem.CreateDirectiveAttribute(
                 displayText,
-                insertText,
+                insertTextSpan.ToString(),
                 descriptionInfo: new([.. attributeDescriptions]),
-                commitCharacters: razorCommitCharacters.ToImmutableAndClear());
+                commitCharacters: razorCommitCharacters.ToImmutableAndClear(),
+                isSnippet);
 
             completionItems.Add(razorCompletionItem);
         }
 
         return completionItems.ToImmutableAndClear();
 
+        bool TryGetSnippetText(
+            RazorSyntaxNode owner,
+            ReadOnlySpan<char> baseTextSpan,
+            RazorCompletionOptions razorCompletionOptions,
+            out ReadOnlySpan<char> snippetTextSpan)
+        {
+            if (razorCompletionOptions.SnippetsSupported
+                // Don't create snippet text when attribute is already in the tag and we are trying to replace it
+                // Otherwise you could have something like @onabort=""=""
+                && containingAttribute is not (MarkupTagHelperDirectiveAttributeSyntax or MarkupAttributeBlockSyntax)
+                && containingAttribute.Parent is not (MarkupTagHelperDirectiveAttributeSyntax or MarkupAttributeBlockSyntax))
+            {
+                var suffixTextSpan = razorCompletionOptions.AutoInsertAttributeQuotes ? "=\"$0\"".AsSpan() : "=$0".AsSpan();
```

Comment:
I think it would be better to store these two spans as `static ReadOnlyMemory<char>` so they're only created once.

Reply:
Learning opportunity for me. What extra allocations happen here? I would have thought the strings would be compiled into consts, and then the AsSpan calls would just create a struct around that.

Reply:
I believe this is the implementation so it seems allocation does happen and caching makes sense

https://github.com/dotnet/runtime/blob/b02b5a026ce4929f62d22a69980db18e8d19f388/src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.cs#L97C42-L97C48

Reply:
... or maybe not, looking at it more closely. Either way, seems like caching can't hurt.